File: array/367_Valid_Perfect_Square/367.cpp
class Solution {
public:
    bool isPerfectSquare(int num) {
        if (num < 2) return true;
        long left = 1, right = num / 2, middle;
        while (left <= right) {
            middle = left + ((right - left) >> 1);
            if (middle * middle < num) {
                left = middle + 1;
            } else if (middle * middle > num) {
                right = middle - 1;
            } else {
                return true;
            }
        }
        return false;
    }
};

File: array/367_Valid_Perfect_Square/367.py
class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        if num < 2:
            return True
        left, right = 1, num // 2
        while left <= right:
            middle = left + ((right - left) >> 1)
            if middle * middle < num:
                left = middle + 1
            elif middle * middle > num:
                right = middle - 1
            else:
                return True
        return False